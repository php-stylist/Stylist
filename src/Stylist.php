<?php declare(strict_types = 1);

namespace Stylist;

use Nette\Utils\Finder;
use Stylist\Checks\CheckInterface;
use Stylist\Fixing\CannotWriteFileException;
use Stylist\Fixing\FileFixer;
use Stylist\Output\OutputInterface;


final class Stylist
{

	/** @var CheckInterface[] */
	private $checks;

	/** @var bool */
	private $onlyCheck = false;

	/** @var OutputInterface */
	private $output;

	/** @var FileFactory */
	private $fileFactory;

	/** @var FileFixer */
	private $fileFixer;

	/** @var string[] */
	private $accepted = [];

	/** @var string[] */
	private $excluded = [];

	/** @var File[] */
	private $files = [];


	/**
	 * @param CheckInterface[] $checks
	 */
	public function __construct(
		array $checks,
		OutputInterface $output,
		FileFactory $fileFactory,
		FileFixer $fileFixer
	)
	{
		$this->checks = $checks;
		$this->output = $output;
		$this->fileFactory = $fileFactory;
		$this->fileFixer = $fileFixer;
	}


	public function accept(array $accepted): self
	{
		$this->accepted = $accepted;
		return $this;
	}


	public function exclude(array $excluded): self
	{
		$this->excluded = $excluded;
		return $this;
	}


	public function onlyCheck(bool $onlyCheck = true): self
	{
		$this->onlyCheck = $onlyCheck;
		return $this;
	}


	/**
	 * @param string[] $paths
	 */
	public function check(array $paths): CheckResult
	{
		\set_time_limit(0);

		$this->output->initialize($paths);

		$start = \microtime(true);
		$success = true;
		$checkedFileNames = [];

		foreach ($this->getFiles($paths) as $file) {
			if (\in_array($file->getPathname(), $checkedFileNames, true)) {
				continue;
			}

			$checkedFileNames[] = $file->getPathname();
			$this->files[] = $checkedFile = $this->fileFactory->create($file);
			$success = $this->checkFile($checkedFile) && $success;
			$this->output->checkedFile($checkedFile);
		}

		$timeTaken = \microtime(true) - $start;
		$result = new CheckResult(
			$success,
			$this->files,
			$timeTaken
		);

		$this->output->finish($result);
		return $result;
	}


	/**
	 * @param string[] $paths
	 * @return \SplFileInfo[]|\Generator
	 */
	private function getFiles(array $paths): \Generator
	{
		foreach ($paths as $path) {
			if (\is_file($path)) {
				yield $path;

			} else {
				yield from Finder::findFiles($this->accepted)
					->exclude($this->excluded)
					->from($path)
					->exclude($this->excluded);
			}
		}
	}


	private function checkFile(File $file): bool
	{
		\set_error_handler(static function ($severity, $message, $file, $line): void {
			throw new \ErrorException($message, 0, $severity, $file, $line);
		});

		try {
			foreach ($this->checks as $check) {
				$check->check($file);
			}

			$this->fixFile($file);
			$file->finishedCheck();

		} catch (\Throwable $error) {
			$file->setCheckError($error);
		}

		\restore_error_handler();
		return $file->isOk();
	}


	private function fixFile(File $file): void
	{
		if ($this->onlyCheck) {
			return;
		}

		try {
			$fixedAll = $this->fileFixer->fixIssues($file);
			if ( ! $fixedAll) {
				$file->addNote(
					'Some of the fixes could not be applied because of conflicting changes. '
					. 'Please run Stylist again to fix them.'
				);
			}

		} catch (CannotWriteFileException $e) {
			$file->addNote('Could not write the file after applying fixes; please make sure that it is writable.');
		}
	}

}
