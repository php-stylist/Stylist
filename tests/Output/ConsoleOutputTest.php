<?php declare(strict_types = 1);

namespace Stylist\Tests\Output;

use Stylist\CheckResult;
use Stylist\File;
use Stylist\IgnoredIssues\IgnoredIssues;
use Stylist\Output\ConsoleOutput;
use Stylist\Tests\DummyCheck;
use Stylist\Tokenista\Tokens;
use Symfony\Component\Console\Output\StreamOutput;
use Tester\Assert;
use Tester\FileMock;
use Tester\TestCase;


require_once __DIR__ . '/../bootstrap.php';


/**
 * @testCase
 */
final class ConsoleOutputTest extends TestCase
{

	public function testOutput(): void
	{
		$stream = FileMock::create('');
		$resource = \fopen($stream, 'w');
		$symfonyOutput = new StreamOutput($resource);

		$console = new ConsoleOutput('/working/directory', $symfonyOutput);
		$console->initialize(['foo']);

		$files = [];
		$tokens = Tokens::from('<?php');

		$files[] = $file1 = new File(new \SplFileInfo('/working/directory/foo/a'), $tokens, [], new IgnoredIssues([]));
		$console->checkedFile($file1);

		$files[] = $file2 = new File(new \SplFileInfo('/working/directory/foo/b'), $tokens, [], new IgnoredIssues([]));
		$file2->addIssue(new DummyCheck(), 'Error', 5);
		$console->checkedFile($file2);

		$files[] = $file3 = new File(new \SplFileInfo('/working/directory/foo/c'), $tokens, [], new IgnoredIssues([]));
		$file3->addIssue(new DummyCheck(), 'Error 1', 4);
		$file3->addIssue(new DummyCheck(), 'Error 2', 12);
		$console->checkedFile($file3);

		$files[] = $file4 = new File(new \SplFileInfo('/working/directory/foo/d'), $tokens, [], new IgnoredIssues([]));
		$file4->addIssue(new DummyCheck(), 'Error 1', 14);
		$file4->addIssue(new DummyCheck(), 'Error 2', 29);
		$file4->addIssue(new DummyCheck(), 'Error 3', 3);
		$console->checkedFile($file4);

		$files[] = $file5 = new File(new \SplFileInfo('/working/directory/foo/e'), $tokens, [], new IgnoredIssues([]));
		$file5->setCheckError(new \LogicException('Serious error during file check!'));
		$console->checkedFile($file5);

		$files[] = $file6 = new File(new \SplFileInfo('/working/directory/foo/f'), $tokens, [], new IgnoredIssues([]));
		$console->checkedFile($file6);

		$result = new CheckResult(false, $files, 0.5);
		$console->finish($result);

		\fclose($resource);

		Assert::matchFile(
			__DIR__ . '/ConsoleOutputTest.expected.txt',
			\trim(\file_get_contents($stream))
		);
	}

}


(new ConsoleOutputTest())->run();
