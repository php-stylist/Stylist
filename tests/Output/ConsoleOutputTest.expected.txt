Stylist v%A% @ PHP %A%

Scanning the following paths:
 - foo

.EEEE.


1 issue found in foo/b
    5     Error

2 issues found in foo/c
    4     Error 1
   12     Error 2

3 issues found in foo/d
    3  âˆš  Error 3
   14     Error 1
   29  x  Error 2

    -     Some of the fixes could not be applied because of conflicting changes.

0 issues found in foo/e
 ERROR    Could not check further due to uncaught LogicException: Serious error during file check!


Checked 6 files in 0.5 seconds.
Finished with issues.
